---
title: "DataCheck"
date: "2023-08-10"
output: html_document
---

## Preface

Theme:  Investigating the identifying omic differences by age and ethnicity on outcomes from mucinous ovarian cancer

Background: 

Ovarian serous cystadenocarcinoma is the **malignant form of ovarian serous tumor**, the most common type of ovarian epithelial tumor. It is the most common type of ovarian malignancy

Data Source:

-   [cbioportoal](https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018) , includes 585 patients/samples, 102 (17.4%) ovarian serous cystadenocarcinoma, 483 serous cystadenocarcinoma (82.6%)
-   [GDC TCGA Ovarian Cancer](https://xenabrowser.net/datapages/?cohort=GDC%20TCGA%20Ovarian%20Cancer%20(OV)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443) : include copy number, DNA Methylation, gene expression, phenotype, somatic mutation level difference.

## Preparation

```{r}
library(tidyverse)
library(dplyr)
library(ggplot2)
```

### cbioportal data

```{r}
ov_pan_sample_df <- read.table("../data/ov_tcga_pan_can_atlas_2018/data_clinical_sample.txt", 
                            header = TRUE, sep = '\t')
ov_pan_sample_df %>% select(TUMOR_TYPE) %>% unique()
head(ov_pan_sample_df)

# Tumor type: Ovarian Serous Cystadenocarcinoma å’Œ Serous Cystadenocarcinoma
# Cancer type is both Ovarian Epithelial Tumor, 
# Cancer type are both Serous ovarian cancer
```

454 Samples have both `ethnicity` and `AGE` information 

```{r}
ov_pan_patient_df <- read.table("../data/ov_tcga_pan_can_atlas_2018/data_clinical_patient.txt", 
                                header = TRUE, sep = "\t")

# na omit
ov_pan_patient_df_omitted <- ov_pan_patient_df[!is.na(ov_pan_patient_df$RACE), ]
# Detect empty strings
empty_rows <- !nzchar(trimws(ov_pan_patient_df_omitted$RACE))
# Keep only rows that are not empty
ov_pan_patient_df_omitted <- ov_pan_patient_df_omitted[!empty_rows, ]
dim(ov_pan_patient_df_omitted) # 454 rows
# head(ov_pan_patient_df_omitted)  
```

Count ethnicity information

```{r}
ov_pan_patient_df_omitted |>
  mutate(Race_Grouped = if_else(RACE == "White", "White", "Other")) |>
  group_by(Race_Grouped) |> 
  summarise(n = n()) 
```

## Mutation & Ethinicity

### Check mutations

Merge patient and sample together 
```{r}
ov_patient_sample_df <- merge(ov_pan_patient_df_omitted, ov_pan_sample_df[,c("PATIENT_ID", "SAMPLE_ID", "TUMOR_TYPE")], by = "PATIENT_ID")
# head(ov_patient_sample_df)  #SAMPLE_ID TCGA-04-1331-01
```

Load mutation data
```{r}
ov_pan_mutations_df <- read.table("../data/ov_tcga_pan_can_atlas_2018/data_mutations.txt", 
                            header = TRUE, sep = "\t")
# head(ov_pan_mutations_df)
```

Merge `mutation` data and `patient` data

```{r}
ov_mutation_meta <- merge(ov_pan_mutations_df, ov_patient_sample_df, by.x = "Tumor_Sample_Barcode", by.y = "SAMPLE_ID")
# head(ov_mutation_meta)
```

Group `RACE` into `White` and `Other` and then summarise the number of each type of variant (DEL, INS, ONP, SNP) in each patient. 

==> Get number of variant per patient. 

```{r}
race_variant_type_count <- ov_mutation_meta %>%  
  mutate(Race_Grouped = if_else(RACE == "White", "White", "Other")) %>% 
  group_by(PATIENT_ID, Variant_Type, Race_Grouped) %>% 
  summarise(n = n()) %>% ungroup()

# head(race_variant_type_count)
```

Plot the number of variant per patient in a box plot. Obviously there is a outlier in `Other - SNP` group. A patient has more than 1500 SNPs.

```{r}
race_variant_type_count |> 
  ggplot(aes(x = Variant_Type, y = n, fill = Race_Grouped)) +
  geom_boxplot(position=position_dodge()) +
  xlab("Variant type") + ylab("Average counts per sample") + 
  guides(fill=guide_legend(title="Race")) +
  theme_bw() + theme(text = element_text(size = 8))
```

Removing that outlier the result showed that `White` group and `Other` seems to have similar distribution. 
```{r}
variant_count_fig <- race_variant_type_count |> 
  filter(n < 500) |> 
  ggplot(aes(x = Variant_Type, y = n, fill = Race_Grouped)) +
  geom_boxplot(position=position_dodge(), outlier.size=.2, linewidth = .2) +
  xlab("Variant type") + ylab("Number of variants per patient") + 
  guides(fill=guide_legend(title="Race")) + facet_wrap(~Variant_Type, scales = "free", ncol = 4) + 
  theme_bw() + theme(text = element_text(size = 8)) 

variant_count_fig

ggsave("./fig_datacheck/00_VariantType_Number.png", plot = variant_count_fig, dpi = 300, height = 60, width = 100, units = "mm")
```

#### Outlier check 

The outlier patient is `TCGA-09-2044` 

```{r}
race_variant_type_count %>% filter(n > 500)
```

Is such a high value generated by many samples from a single person? 

> No. This person only contribute one sample to the study 

```{r}
ov_pan_sample_df %>% filter(PATIENT_ID == "TCGA-09-2044")
```
### Summary in detail 

Whether `Other` group has a higher SNP per patient in detail? Which type of SNP in detail?

```{r}
# head(ov_mutation_meta[,c("Consequence", "Variant_Classification", "Variant_Type")], 100) 
```

```{r}
# In total, there are 57 different consequence SNP
ov_mutation_meta |> 
  select(Consequence) |> 
  unique()
```

```{r}
# In total, there aer 1859 different Variant Type, but many of them are SNP novel + gene , thus outnumbered 
mutation_summary <- ov_mutation_meta |>
  select(Variant_Type, Variant_Classification) |> 
  group_by(Variant_Type, Variant_Classification) |>
  summarise(n = n()) |> 
  ungroup()

mutation_summary
```


```{r include=FALSE}
# SNP type = Novel + RS
# 1416 Novel
ov_mutation_meta |>
  select(Variant_Type, Variant_Classification) |> 
  filter(str_detect(Variant_Classification, "novel")) |>
  summarise(n = n()) |> 
  ungroup()

# 835 rs -> means the rs number in dbSNP 
ov_mutation_meta |>
  filter(str_detect(Variant_Classification, "rs")) |> 
  summarise(n = n()) |> 
  ungroup()

# Still there are 24 SNP records, there dsSNP_RS status is either rs... or novel but Variant_Classification field does not contain this information
ov_mutation_meta |>
  filter(!str_detect(Variant_Classification, "rs")) |> 
  filter(!str_detect(Variant_Classification, "novel")) |> 
  filter(str_detect(Variant_Classification, "TCGA"))
```

Modify `ov_mutation_classification` focusing on `novel` and `rs` 

```{r}
ov_mutation_meta_short <- ov_mutation_meta |>
  mutate(Variant_Classification_Short = case_when(
    str_detect(dbSNP_RS, "novel") ~ "novel",
    str_detect(dbSNP_RS, "rs") ~ "known",
    str_detect(Variant_Classification, "novel") ~ "novel",
    str_detect(Variant_Classification, "rs") ~ "known",
    TRUE ~ Variant_Classification
  ))

dim(ov_mutation_meta_short) # 19561 

# head(ov_mutation_meta_short, 20)
```

There are 15 types in total, the top 5 groups are "novel", "known", "Missense_Mutation", "Silent", and "Nonsense_Mutation"

Summarize the number of variants in each group. 

```{r}

ov_mutation_meta_short |> 
  mutate(Race_Grouped = if_else(RACE == "White", "White", "Other")) |> 
  group_by(Race_Grouped, Variant_Classification_Short) |>
  summarise(n = n())  |>
  ungroup() |> pivot_wider(names_from = "Race_Grouped", values_from = "n") |> 
  mutate(sum = Other + White) |> arrange(desc(sum))

facet_short_order <- ov_mutation_meta_short |> 
  group_by(Variant_Classification_Short) |>
  summarise(n = n())  |>
  ungroup() |>
  arrange(desc(n)) |> select(Variant_Classification_Short)

facet_short_order$Variant_Classification_Short
```

Check the top 5 groups including the outlier point. This outlier point biased the distribution in `known` and `novel` group. 

```{r}
desired_variants <- c("novel", "known", "Missense_Mutation", "Silent", "Nonsense_Mutation")

Variant_type_detail_fig <- ov_mutation_meta_short |> 
  # create new column
  mutate(Variant_merged = if_else(Variant_Classification_Short %in% desired_variants, Variant_Classification_Short, "Others"),
        Race_Grouped = if_else(RACE == "White", "White", "Other")) |> 
  
  group_by(PATIENT_ID, Race_Grouped, Variant_Type, Variant_merged) |> 
  summarise(n = n()) %>% replace_na(list(n = 0)) |> 
  ungroup() |>
  ggplot(aes(x = Variant_Type, y = n, fill = Race_Grouped)) +
  geom_boxplot(position=position_dodge(), width = 0.5, outlier.size=.2, linewidth = .2) + 
  xlab("Variant type") + ylab("Number of variants per patient") + 
  theme_bw() + theme(text = element_text(size = 10)) + 
  guides(fill=guide_legend(title="Race")) +
  facet_wrap(~Variant_merged, scales = "free")

Variant_type_detail_fig
```

Filtering out the outlier 
```{r}
Variant_type_detail_fig <- ov_mutation_meta_short |> 
  # create new column
  mutate(Variant_merged = if_else(Variant_Classification_Short %in% desired_variants, Variant_Classification_Short, "Others"),
        Race_Grouped = if_else(RACE == "White", "White", "Other")) |> 
  
  group_by(PATIENT_ID, Race_Grouped, Variant_Type, Variant_merged) |> 
  filter(PATIENT_ID != "TCGA-09-2044") |>
  summarise(n = n()) %>% replace_na(list(n = 0)) |> 
  ungroup() |>
  ggplot(aes(x = Variant_Type, y = n, fill = Race_Grouped)) +
  geom_boxplot(position=position_dodge(), width = 0.5, outlier.size=.2, linewidth = .2) + 
  xlab("Variant type") + ylab("Number of variants per patient") + 
  theme_bw() + theme(text = element_text(size = 10)) + 
  guides(fill=guide_legend(title="Race")) +
  facet_wrap(~Variant_merged, scales = "free")

Variant_type_detail_fig

ggsave("./fig_datacheck/00_VariantType_Number_detail.png", plot = Variant_type_detail_fig, dpi = 300, height = 80, width = 140, units = "mm")
```



```{r}

variant_type_detail_full <- ov_mutation_meta_short |> 
  # create new column
  mutate(Variant_merged = if_else(Variant_Classification_Short %in% desired_variants, Variant_Classification_Short, "Others"),
        Race_Grouped = if_else(RACE == "White", "White", "Other")) |> 
  
  group_by(PATIENT_ID, Race_Grouped, Variant_Type, Variant_Classification_Short) |> 
  filter(PATIENT_ID != "TCGA-09-2044") |>
  summarise(n = n()) %>%
  ungroup() 

variant_type_detail_full$Variant_Classification_Short <- factor(variant_type_detail_full$Variant_Classification_Short, levels = facet_short_order$Variant_Classification_Short)

Variant_type_detail_fig <- variant_type_detail_full |> 
  ggplot(aes(x = Variant_Type, y = n, fill = Race_Grouped)) +
  geom_boxplot(position=position_dodge(), width = 0.5, outlier.size=.2, linewidth = .2) + 
  xlab("Variant type") + ylab("Number of variants per patient") +
  guides(fill=guide_legend(title="Race")) +
  facet_wrap(~Variant_Classification_Short, scales = "free", ncol = 3) +
  theme_bw() + theme(text = element_text(size = 8)) + theme(strip.text.x = element_text(size = 8))

Variant_type_detail_fig

ggsave("./fig_datacheck/00_VariantType_Number_full.png", plot = Variant_type_detail_fig, dpi = 300, height = 160, width = 160, units = "mm")
```

## Mutation / ethinicity / survial

### is ethinicity related to survival?

#### Load data 

Survival data is from _GDC TCGA Ovarian Cancer_

OS (Overall survival), refers to the primary endpoint in many cancer studies. Overval survival measures the duration of time from either the data of diagnosis or the start of treatment for a disease that a patient is still alive. 

```{r}
survival_df <- read.table("../data/ov_tcga_GDC/TCGA-OV.survival.tsv", header = TRUE, sep = "\t")
head(survival_df, 20)

survival_df %>% 
  mutate(OS = factor(OS)) %>% group_by(OS) %>% 
  summarise(n = n())

# in total 
# 296 are 0 -> which means alive or dead tumor free
# 435 are 1:DEAD WITH TUMOR
```

#### Merge with prefiltered data

Some patient have multiple samples such as "TCGA-04-1337" have 	"TCGA-04-1337-11A", "TCGA-04-1337-01A".

```{r}
ov_survival_meta <- merge(survival_df, ov_patient_sample_df[c("AGE", "PATIENT_ID", "RACE")], by.x = "X_PATIENT", by.y = "PATIENT_ID")
dim(ov_survival_meta)
head(ov_survival_meta)
```

To remove redundancy, only the record with the highest OS.time was kept. This operation left 309 rows in total. 

```{r}
# in total 454 data were kept 
ov_survival_meta %>% 
  select(!sample) %>% 
  group_by(X_PATIENT) %>%
  filter(OS.time == max(OS.time)) %>%
  ungroup() %>% unique()
```

Among these 309 records, 273 were "White" and 36 were "Other" 

```{r}
ov_survival_meta %>% 
  select(!sample) %>% 
  group_by(X_PATIENT) %>%
  filter(OS.time == max(OS.time)) %>%
  ungroup() %>% unique() %>%
  mutate(Race_grouped = ifelse(RACE == "White", "White", "Other")) %>% 
  group_by(Race_grouped) %>% 
  summarise(n = n())
```

Whether different ethnicities differ in terms of OS or OS.time?

```{r}
# install.packages('survival')
library(survival)
library(ggsignif)

patient_OS_time_AGE_RACE <- ov_survival_meta %>% 
  select(!sample) %>% 
  group_by(X_PATIENT) %>%
  filter(OS.time == max(OS.time)) %>%
  ungroup() %>% unique() %>%
  mutate(Race_grouped = ifelse(RACE == "White", "White", "Other"))

Race_OStime_boxfig <- patient_OS_time_AGE_RACE %>% 
  ggplot(aes(x = Race_grouped, y = OS.time)) + 
  geom_boxplot(width = 0.5) + 
  xlab("Race") + 
  theme_bw() + theme(text = element_text(size = 8))

Race_OStime_boxfig
ggsave("./fig_datacheck/00_RaceOStime_box.png", plot = Race_OStime_boxfig, dpi = 300, height = 50, width = 50, units = "mm")

Race_OStime_density <- patient_OS_time_AGE_RACE %>% 
  ggplot(aes(x = OS.time, fill = Race_grouped)) + 
  geom_density(alpha = 0.5) + 
  guides(fill=guide_legend(title="Race")) + 
  theme_bw() + theme(text = element_text(size = 8))
Race_OStime_density
ggsave("./fig_datacheck/00_RaceOStime_density.png", plot = Race_OStime_density, dpi = 300, height = 50, width = 80, units = "mm")

```

## Survival analysis (KM)


Create survival objective using `Surv` to combine survival time and event indicator.

```{r}
# install.packages('survival')
# install.packages('ggsurvfit')
# install.packages(c("lubridate", "ggsurvfit", "gtsummary", "tidycmprsk"))
library(survival)
library(lubridate)
library(ggsurvfit)
library(gtsummary)
library(tidycmprsk)


surv_obj <- Surv(patient_OS_time_AGE_RACE$OS.time, patient_OS_time_AGE_RACE$OS)
str(surv_obj)
```

```{r}
survfit2(Surv(OS.time, OS) ~ 1, data = patient_OS_time_AGE_RACE) %>% 
  ggsurvfit() + 
  labs(
    x = "Days",
    y = "Overvall survival proabbility"
  ) +  add_confidence_interval() 
```

```{r}
survfit_fig <- survfit2(Surv(OS.time, OS) ~ Race_grouped, data = patient_OS_time_AGE_RACE) %>% 
  ggsurvfit(theme = theme_ggsurvfit_default()) + 
  labs(
    x = "Days",
    y = "Overvall survival proabbility"
  ) +  add_confidence_interval() 

survfit_fig

ggsave("./fig_datacheck/00_surfit_fig.png", plot = survfit_fig, dpi = 300, height = 80, width = 100, units = "mm")

```


Use `logrank` test to test whether different factors have significant difference
Since "AGE" is another putative varaible related to the "OS.time".Thus, use `strata` to include it.
The following result gave a p-value = 0.007 < 0.05, meaning that the overval survival probability is significantly different between group "White" and group "Other".

```{r}
surv_obj <- Surv(patient_OS_time_AGE_RACE$OS.time, patient_OS_time_AGE_RACE$OS)
survdiff(surv_obj ~ patient_OS_time_AGE_RACE$Race_grouped + strata(patient_OS_time_AGE_RACE$AGE))
```








